### PR TITLE
Skip cythonizing with no changes

### PR DESCRIPTION
**Pull request by wkentaro**
To make it developer friendly.

### CODE REVIEW

**Comment by okuta** [comment]
I recommend using the `ccache`. You can reduce the compile time safety.

**Comment by wkentaro** [comment]
Thanks for the comment, but skipping compilation is still much more faster, isn't it?

On OS X Yosemite, ccache setup
```
brew install ccache
sudo ln -s /usr/local/bin/ccache /usr/local/bin/clang
sudo ln -s /usr/local/bin/ccache /usr/local/bin/clang++
```


1 without ccache
```
Using /usr/local/lib/python2.7/site-packages
Finished processing dependencies for chainer==1.6.2
python setup.py develop  33.12s user 2.20s system 97% cpu 36.118 total
```

2 with ccache
```
Using /usr/local/lib/python2.7/site-packages
Finished processing dependencies for chainer==1.6.2
python setup.py develop  8.25s user 1.43s system 97% cpu 9.967 total
```

3 skipping with no change
```
Using /usr/local/lib/python2.7/site-packages
Finished processing dependencies for chainer==1.6.2
python setup.py develop  0.82s user 0.44s system 95% cpu 1.330 total
```

**Comment by okuta** [comment]
Thank you for the presentation of evidence.
It is necessary to check environment variables and header files that in order to determine the skipping compilation.
We need a more safety method.
For example, to control this function by command line flag.

**Comment by wkentaro** [comment]
> It is necessary to check environment variables and header files that in order to determine the skipping compilation.
We need a more safety method.

I see, that's reasonable. I added a command line option `--cupy-only-changed` to skip cythonizing if there're no changes.

File: chainer_setup_build.py [hunk 1]
```diff
@@ -388,6 +421,10 @@ def cythonize(extensions, force=False, annotate=False, compiler_directives={}):
 
     cython_cmdbase.extend(['--fast-fail', '--verbose', '--cplus'])
     for ext in extensions:
+        if all(not _is_changed(f) for f in ext.sources) and only_changed:
```
Comment (okuta) [inline chainer_setup_build.py]: How about `if only_changed and all(not _is_changed(f) for f in ext.sources):`?

Comment (wkentaro) [inline chainer_setup_build.py]: I need to call `_is_changed` function to update the `*.pyx.md5` file.

Comment (wkentaro) [inline chainer_setup_build.py]: I added a comment to describe that, looks good?

Comment (okuta) [inline chainer_setup_build.py]: I think that the calculation of md5 is not necessary for most users.
So, it is better that extra file is not made.
If you want to check the md5 of all files, following code is good.
`if all([not _is_changed(f) for f in ext.sources]) and only_changed:`

Comment (wkentaro) [inline chainer_setup_build.py]: What's the difference between 
`if all(not _is_changed(f) for f in ext.sources) and only_changed:`
and
`if all([not _is_changed(f) for f in ext.sources]) and only_changed:`
?

Comment (okuta) [inline chainer_setup_build.py]: `not _is_changed(f) for f in ext.sources` return the generator.
And, `all` method performs short-circuit evaluation.
`[not _is_changed(f) for f in ext.sources]` return the list.
`_is_changed` is called for all elements of `ext.sources`.

Comment (wkentaro) [inline chainer_setup_build.py]: Ah, I understood. Thanks.

Comment (wkentaro) [inline chainer_setup_build.py]: I updated the commit.

**Comment by okuta** [comment]
I think that the calculation of md5 is not necessary for most users.
So, it is better that extra file is not made.

**Comment by wkentaro** [comment]
Ok. In that case, maybe there is no way to fix this issue.
And at the current master branch, the speed is fast enough, so this PR can be closed.

**Pull request closed without merging**